### PR TITLE
Broadcasting

### PR DESCRIPTION
**Pull request by kaushikcfd**

### CODE REVIEW

File: pytato/array.py [hunk 1]
```diff
@@ -585,30 +585,33 @@ def _binary_op(self,
...
-        def add_indices(val: prim.Expression) -> prim.Expression:
-            if self.ndim == 0:
+        def add_indices(val: prim.Expression,
+                        s: ShapeType, r: ShapeType) -> prim.Expression:
+            if len(s) == 0:
                 return val
             else:
-                indices = tuple(var(f"_{i}") for i in range(self.ndim))
+                indices = utils.get_indexing_expression(s, r)
                 return val[indices]
```
Comment (inducer) [inline pytato/array.py:596]: Could/should be moved into `utils`.

File: pytato/utils.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+            result.append(dim1)
+        else:
+            raise ValueError("operands could not be broadcast together with shapes "
+                             f"{s1} {s2}.")
+
+    return tuple(result)
+
+
+def get_indexing_expression(s: ShapeType,
+                            r: ShapeType) -> Tuple[ScalarExpression, ...]:
```
Comment (inducer) [inline pytato/utils.py:52]:
```suggestion
def get_indexing_expression(shape: ShapeType,
                            result_shape: ShapeType) -> Tuple[ScalarExpression, ...]:
```

File: pytato/array.py [hunk 3]
```diff
@@ -585,30 +585,33 @@ def _binary_op(self,
             other: Union[Array, Number],
             get_result_type: Callable[[DtypeOrScalar, DtypeOrScalar], np.dtype[Any]] = np.result_type,  # noqa
             reverse: bool = False) -> Array:
+        import pytato.utils as utils
 
-        def add_indices(val: prim.Expression) -> prim.Expression:
-            if self.ndim == 0:
+        def add_indices(val: prim.Expression,
+                        s: ShapeType, r: ShapeType) -> prim.Expression:
```
Comment (inducer) [inline pytato/array.py:591]:
```suggestion
                        shape: ShapeType, result_shape: ShapeType) -> prim.Expression:
```
(not wild about `s` and `r` as argument names)

File: pytato/utils.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+from typing import Tuple
+from pytato.array import ShapeType
+from pytato.scalar_expr import ScalarExpression
+
+
+def get_shape_after_broadcasting(s1: ShapeType, s2: ShapeType) -> ShapeType:
+    s_big, s_small = (s1, s2) if len(s1) > len(s2) else (s2, s1)
+
+    result = list(s_big[:len(s_big)-len(s_small)])
```
Comment (inducer) [inline pytato/utils.py:35]:
```suggestion
    # Use leading dimensions from s_big unmodified.
    result = list(s_big[:len(s_big)-len(s_small)])
```

File: test/test_codegen.py [hunk 5]
```diff
@@ -518,6 +518,28 @@ def test_passsing_bound_arguments_raises(ctx_factory):
         evt, (out2,) = prg(queue, x=np.random.rand(10))
 
 
+@pytest.mark.parametrize("shape1, shape2", (
+                                            [(10, 4), ()],
+                                            [(), (10, 4)],
+                                            [(3,), (32, 32, 3)],
+                                            [(32, 32, 3), (3,)],
+                                            [(4, 1, 3), (1, 7, 1)],
```
Comment (inducer) [inline test/test_codegen.py:526]: Add tests for varying lengths including one-long dimensions.

**Review by inducer — COMMENTED** [review]
Thanks! Just a few things here.

File: test/test_codegen.py (hunk 5 above)
Comment (kaushikcfd) [reply test/test_codegen.py:526]: By varying do you mean a parametric axis length?

Comment (inducer) [reply test/test_codegen.py:526]: I didn't think that (but maybe that's useful if easy?). What I meant was to include examples that contain one-long dimensions where `len()` of both shapes isn't the same.

**Comment by inducer** [comment]
Unsubscribing... @-mention or request review once it's ready for a look or needs attention.

File: test/test_codegen.py (hunk 5 above)
Comment (kaushikcfd) [reply test/test_codegen.py:527]: e13e5b3

Comment (kaushikcfd) [reply test/test_codegen.py:527]: Let me know if this is not what you had in mind.

File: pytato/array.py (hunk 1 above)
Comment (kaushikcfd) [reply pytato/array.py]: fe0852e

File: pytato/array.py [hunk 6]
```diff
@@ -585,30 +585,33 @@ def _binary_op(self,
...
+            if len(s) == 0:
                 return val
             else:
-                indices = tuple(var(f"_{i}") for i in range(self.ndim))
+                indices = utils.get_indexing_expression(s, r)
                 return val[indices]
 
         if isinstance(other, Array):
             if self.namespace is not other.namespace:
                 raise ValueError("Operands must belong to the same namespace.")
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply pytato/array.py:595]: fe0852e

**Comment by kaushikcfd** [comment]
@inducer: Thanks for the review, this is in a better state now.

File: test/test_codegen.py (hunk 5 above)
Comment (inducer) [reply test/test_codegen.py:528]: This works, thanks!

**Review by inducer — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**